processing/src/main/java/io/druid/query/lookup/LookupExtractionFn.java

```diff
@@ -90,7 +90,8 @@ public boolean isInjective()
   @JsonProperty("optimize")
   public boolean isOptimize()
   {
-    return optimize;
+    // case retainMissingValue == true  we can not optimize need to push down the filter
+    return !isRetainMissingValue() && optimize;
```

Comment:
to elaborate more, maybe it's possible to have some optimization with retainMissingValues=true in the future in some other use case, the currently incompatibility of retainMissingValues and optimize() is specific to the ExtractionDimFilter optimization being used, so the check should go in the filter that's affected